service/history/queues/reader_quotas.go

```diff
@@ -0,0 +1,86 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"strconv"
+
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/quotas"
+)
+
+const (
+	readerRequestToken = 1
+)
+
+func NewReaderPriorityRateLimiter(
+	rateFn quotas.RateFn,
+	maxReaders int,
+) quotas.RequestRateLimiter {
+	rateLimiters := make(map[int]quotas.RateLimiter, maxReaders)
+	readerCallerToPriority := make(map[string]int, maxReaders)
+	for readerId := defaultReaderId; readerId != defaultReaderId+maxReaders; readerId++ {
+		// use readerId as priority
+		rateLimiters[readerId] = quotas.NewDefaultOutgoingRateLimiter(rateFn)
+		readerCallerToPriority[newReaderRequest(int32(readerId)).Caller] = readerId
```

Comment:
Took me a while to understand why we need newReaderRequest(int32(readerId)).Caller here.
We could add comment here that each reader will use ReaderID as Caller when using this rate limiter.